Add ContainsKeys utility function

bharatnc opened the pull request:
This PR adds a convenient utility function called `ContainsKeys`.

The function takes in a map of labels (key and val) of a runtime object and a list of label (keys) and returns true if all of the keys in the given list are present in the map of labels and returns false otherwise.

This PR also leverage the use of this new function in the following files:

https://github.com/elastic/cloud-on-k8s/blob/master/pkg/controller/kibanaassociation/labels.go#L33
https://github.com/elastic/cloud-on-k8s/blob/master/pkg/controller/elasticsearch/remotecluster/remoteca/watches.go#L60

Fixes #2646

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

bharatnc commented:
>  Author of the following commits did not sign a [Contributor Agreement](https://www.elastic.co/contributor-agreement):
> [e6366f1](https://github.com/elastic/cloud-on-k8s/commit/e6366f120ae29c1da6639357bcf1f44f90d392ae)
> 
> Please, read and sign the above mentioned agreement if you want to contribute to this project

Just did sign this.

bharatnc commented:
cc: @barkbay for reviews. Thank you!

bharatnc commented on pkg/utils/label/label_test.go:
```diff
@@ -0,0 +1,66 @@
+func TestHasLabel(t *testing.T) {
+	tests := []struct {
+		name   string
+		inp    *appsv1.Deployment
```
Yes this can be done. Also, the `HasLabel` func should then take `metav1.Object` instead of `runtime.Object`.  `metav1.Object` doesn't implement  all the methods of`runtime.Object` so they're incompatible types.

bharatnc replied:
I've made the change. Thanks!

bharatnc commented on pkg/utils/label/label_test.go:
```diff
@@ -0,0 +1,66 @@
+				},
+			},
+			labels: []string{},
+			want:   false,
```
fixed.

bharatnc commented on pkg/utils/label/label.go:
```diff
@@ -0,0 +1,26 @@
+package labels
```
Added, thank you!

bharatnc commented on pkg/utils/label/label_test.go:
```diff
@@ -0,0 +1,66 @@
+package labels
```
Added.

bharatnc commented on pkg/utils/label/label_test.go:
```diff
@@ -0,0 +1,66 @@
+		want   bool
+	}{
+		{
+			name: "no labels on object",
```
Sure, added the test.

bharatnc commented:
> Thanks for the PR ! 
> 
> I left a few comments. Please add the license header at the beginning of the files _(just copy/paste them from an existing one)_.
> Also we should use this new function where it's useful:
> 
> * https://github.com/elastic/cloud-on-k8s/blob/e7b9da9212c10ad45ff39d3ea9b2ea897973d233/pkg/controller/kibanaassociation/labels.go#L33
> * https://github.com/elastic/cloud-on-k8s/blob/e7b9da9212c10ad45ff39d3ea9b2ea897973d233/pkg/controller/elasticsearch/remotecluster/remoteca/watches.go#L60

@barkbay thank you! Fixed all but one of the suggestions -- I will make changes to use this label utility in the suggested files and update the PR.

bharatnc commented:
> Thanks for the PR ! 
> 
> I left a few comments. Please add the license header at the beginning of the files _(just copy/paste them from an existing one)_.
> Also we should use this new function where it's useful:
> 
> * https://github.com/elastic/cloud-on-k8s/blob/e7b9da9212c10ad45ff39d3ea9b2ea897973d233/pkg/controller/kibanaassociation/labels.go#L33
> * https://github.com/elastic/cloud-on-k8s/blob/e7b9da9212c10ad45ff39d3ea9b2ea897973d233/pkg/controller/elasticsearch/remotecluster/remoteca/watches.go#L60

@barkbay  thank you! I have addressed all the comments. I've also leveraged this utility in both the files that you've mentioned. 

Let me know if if this looks good you.

barkbay commented:
I think the idea is to replace this kind of code:

```golang
		clusterAssociationName, ok := labels[RemoteClusterNameLabelName]
		if !ok {
			return nil
		}
		clusterAssociationNamespace, ok := labels[RemoteClusterNamespaceLabelName]
		if !ok {
			return nil
		}
		return []reconcile.Request{
			{NamespacedName: types.NamespacedName{
				Namespace: clusterAssociationNamespace,
				Name:      clusterAssociationName},
			},
		}
```

with

```golang
		if !maps.ContainsKeys(labels, RemoteClusterNameLabelName, RemoteClusterNamespaceLabelName) {
			return nil
		}
		return []reconcile.Request{
			{NamespacedName: types.NamespacedName{
				Namespace: labels[RemoteClusterNameLabelName],
				Name:      labels[RemoteClusterNamespaceLabelName]},
			},
		}
```

On second thought I'm not sure we need to create something specific to the K8S api. Maybe a simpler solution would be to add a utility function to the existing `maps.go`: 

```golang
func ContainsKeys(m map[string]string, labels ...string) bool {
	for _, key := range m {
		if _, exists := m[key]; !exists {
			return false
		}
	}
	return true
}
```

Sorry for the wrong ideas.

bharatnc commented:
> I think the idea is to replace this kind of code:
> 
> ```go
> 		clusterAssociationName, ok := labels[RemoteClusterNameLabelName]
> 		if !ok {
> 			return nil
> 		}
> 		clusterAssociationNamespace, ok := labels[RemoteClusterNamespaceLabelName]
> 		if !ok {
> 			return nil
> 		}
> 		return []reconcile.Request{
> 			{NamespacedName: types.NamespacedName{
> 				Namespace: clusterAssociationNamespace,
> 				Name:      clusterAssociationName},
> 			},
> 		}
> ```
> 
> with
> 
> ```go
> 		if !maps.ContainsKeys(labels, RemoteClusterNameLabelName, RemoteClusterNamespaceLabelName) {
> 			return nil
> 		}
> 		return []reconcile.Request{
> 			{NamespacedName: types.NamespacedName{
> 				Namespace: labels[RemoteClusterNameLabelName],
> 				Name:      labels[RemoteClusterNamespaceLabelName]},
> 			},
> 		}
> ```
> 
> On second thought I'm not sure we need to create something specific to the K8S api. Maybe a simpler solution would be to add a utility function to the existing `maps.go`:
> 
> ```go
> func ContainsKeys(m map[string]string, labels ...string) bool {
> 	for _, key := range m {
> 		if _, exists := m[key]; !exists {
> 			return false
> 		}
> 	}
> 	return true
> }
> ```
> 
> Sorry for the wrong ideas.

@barkbay Pardon, I think I made a mistake while make the existing code leverage my new utility func. Thanks for your explanation. 

> On second thought I'm not sure we need to create something specific to the K8S api

Sure, I can just implement this as a method inside `maps.go`.

I'll update the PR. Thanks!

bharatnc commented on pkg/controller/elasticsearch/remotecluster/remoteca/watches.go:
```diff
@@ -57,21 +59,16 @@ func AddWatches(c controller.Controller, r *ReconcileRemoteCa) error {
-		}
-		clusterAssociationNamespace, ok := labels[RemoteClusterNamespaceLabelName]
-		if !ok {
+		if labels[common.TypeLabelName] != TypeLabelValue {
```
I think we should still check if the values match.

bharatnc commented on pkg/controller/kibanaassociation/labels.go:
```diff
@@ -30,10 +32,10 @@ func NewResourceSelector(name string) client.MatchingLabels {
 		return false
 	}
-	if ns, ok := labels[AssociationLabelNamespace]; !ok || ns != kibana.Namespace {
+	if labels[AssociationLabelName] != kibana.Name || labels[AssociationLabelNamespace] != kibana.Namespace {
```
Sane here, I think we should still check if the values match.

bharatnc commented:
@barkbay just made those changes, let me know if they look better. And thanks for the review!

barkbay commented:
jenkins test this please

barkbay commented:
jenkins test this please

barkbay commented on pkg/controller/kibanaassociation/labels.go:
```diff
@@ -5,13 +5,15 @@
 	kbv1 "github.com/elastic/cloud-on-k8s/pkg/apis/kibana/v1"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/common"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/common/user"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/label"
```
You should remove this line too

bharatnc replied:
thanks, I missed it. Removed.

barkbay commented on pkg/controller/elasticsearch/remotecluster/remoteca/watches.go:
```diff
@@ -8,10 +8,12 @@ import (
 	esv1 "github.com/elastic/cloud-on-k8s/pkg/apis/elasticsearch/v1"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/common/watches"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/certificates/transport"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/label"
```
You should remove this line

bharatnc replied:
thanks, I missed it. Removed.

bharatnc commented:
jenkins test this please

barkbay commented:
jenkins test this please

bharatnc commented:
> jenkins test this please

builds are now all green! :+1:

pebrc commented:
Let's maybe update the title and PR description and link to the corresponding issue before we merge this.

bharatnc commented:
> Let's maybe update the title and PR description and link to the corresponding issue before we merge this.

@pebrc  and @barkbay just updated the  PR's title and description and also linked the issue. Let me know if that looks good. Thanks!

bharatnc commented:
@barkbay  @pebrc does this PR looks good to be merged? Thanks!
